[Perf] Cache key size in CngKey

GrabYourPitchforks opened the pull request:
An internal partner is reporting that they're seeing CNG signature validation latency show up as a bottleneck. The culprit seems to be that each signing \& validation operation queries the `CngKey.KeySize` property.

https://github.com/dotnet/runtime/blob/5d7b16f3ccc2680dde0270b2190ab4adb0f813b7/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RSACng.Key.cs#L20-L26

https://github.com/dotnet/runtime/blob/5d7b16f3ccc2680dde0270b2190ab4adb0f813b7/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngAlgorithmCore.cs#L47-L56

The `CngKey.KeySize` property is not cached anywhere in the system, which means that each query to this property results in an LRPC call back into lsass or lsaiso. This context switch is what appears to be introducing the latency.

Since `CngKey` instances can't change their key size once created, we can cache the key size on our end without requiring calls back through the `NCrypt*` (lsass) layer. That significantly reduces the latency.

```cs
public class CngRunner
{
    private RSACng _rsa;
    private ECDsaCng _ecdsa;

    private static byte[] _emptyDigest = new byte[256 / 8];
    private byte[] _rsaSignedHash;
    private byte[] _ecdsaSignedHash;

    [GlobalSetup]
    public void Setup()
    {
        _rsa = new RSACng(2048);
        _ecdsa = new ECDsaCng(256);

        _rsaSignedHash = _rsa.SignHash(_emptyDigest, HashAlgorithmName.SHA256, RSASignaturePadding.Pss);
        if (!Rsa_VerifyHash()) { throw new CryptographicException("Self-test failed."); }

        _ecdsaSignedHash = _ecdsa.SignHash(_emptyDigest);
        if (!Ecdsa_VerifyHash()) { throw new CryptographicException("Self-test failed."); }
    }

    [Benchmark]
    public bool Rsa_VerifyHash() => _rsa.VerifyHash(_emptyDigest, _rsaSignedHash, HashAlgorithmName.SHA256, RSASignaturePadding.Pss);

    [Benchmark]
    public bool Ecdsa_VerifyHash() => _ecdsa.VerifyHash(_emptyDigest, _ecdsaSignedHash);
}
```

|           Method |        Job |    Toolchain |      Mean |    Error |   StdDev | Ratio |
|----------------- |----------- |------------- |----------:|---------:|---------:|------:|
|   Rsa_VerifyHash | Job-EBVYOU | cngkey_cache |  48.03 μs | 0.140 μs | 0.131 μs |  0.38 |
|   Rsa_VerifyHash | Job-QBWHWJ |         main | 125.42 μs | 0.505 μs | 0.394 μs |  1.00 |
|                  |            |              |           |          |          |       |
| Ecdsa_VerifyHash | Job-EBVYOU | cngkey_cache | 220.13 μs | 2.552 μs | 2.731 μs |  0.75 |
| Ecdsa_VerifyHash | Job-QBWHWJ |         main | 295.14 μs | 0.885 μs | 0.739 μs |  1.00 |

### Open discussion

Technically this is a behavioral change, since it means that the `CngKey.KeySize` property will remain accessible on an otherwise disposed instance, just as long as it has already been fetched at least once prior to disposal. I could easily add a disposed check (and unit test to validate this behavior) if we think preserving the earlier behavior is worthwhile.

GrabYourPitchforks committed:
Cache CngKey.KeySize

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs:
```diff
@@ -215,6 +224,7 @@ public int KeySize
                     }
                 }
 
+                _cachedKeySize = keySize;
```
Lines 218-223 have early returns. Do we want to change the switch to assign `keySize`? Otherwise the Platform Crypto Provider won't benefit from the cache.

bartonjs replied:
Good catch.  I remember looking exactly at that and thinking it needed something special, but somehow the flow there caught me as just assigning the local... which, of course, isn't what it does.

GrabYourPitchforks committed:
PR feedback, add tests

GrabYourPitchforks commented:
Updated the PR to use -1 instead of 0 as a sentinel, plus moved some logic around to account for early return scenarios. I also added a unit test to ensure that all non-handle properties throw after the CngKey instance is disposed, even if the properties had been previously accessed and cached. This should help thwart behavioral regressions.

Note to reviewers: I suggest **hiding whitespace changes** when reviewing the latest iteration, since some indentation changed within the KeySize property getter.

Perf results from this build are nearly identical to the earlier run. I won't bother posting them.

vcsjones commented on src/libraries/System.Security.Cryptography.Cng/tests/CngKeyTests.cs:
```diff
@@ -0,0 +1,52 @@
+            }
+            catch
+            {
+                throw new SkipException($"Property getter CngKey.{propertyName} threw an exception; skipping test.");
```
SkipException only works if the `Theory` attribute is `[ConditionalTheory]`. Otherwise the test will just throw the exception and fail.

Why would we want to skip the test though?

GrabYourPitchforks replied:
Because if the property setter doesn't work for some reason unrelated to disposal, that's not really in the purview of this test class.

vcsjones commented on src/libraries/System.Security.Cryptography.Cng/tests/CngKeyTests.cs:
```diff
@@ -0,0 +1,52 @@
+
+namespace System.Security.Cryptography.Tests
+{
+    public static class CngTests
```
Nit: Class name should probably be `CngKeyTests`.

GrabYourPitchforks replied:
Copy and paste error on my part. Doh. :(

brianrob commented:
A bit of justification here: This isn't a big deal for one off usages, but any time there is a system that is caching an instance for use over and over, the LRPC call is a significant bottleneck, and limits the scale of usage to be what lsass can handle.  This is a common situation in cloud services, so a change to cache the key size is worthwile.

vcsjones commented:
@brianrob out of curiosity, are you using RSACng directly? Or are you using RSA.Create?

If the former, are you using CNG specific features that would exclude using RSA.Create?

brianrob commented:
@vcsjones, I'm one of the folks that was analyzing perf data, but don't work on the service itself.  I've sent mail asking them to reply here.

vcsjones commented:
Gotcha. I only ask because starting in .NET 8, `RSA.Create()` will use bcrypt, not ncrypt, when possible. So all the RPC overhead goes away entirely.

GrabYourPitchforks commented:
Internal partner reports that they're using `RSACertificateExtensions.GetRSAPublicKey` extension method. On net8, this should go through bcrypt instead of ncrypt, so you're correct that the perf overhead should disappear.

I've changed the benchmark to perform signing instead of verification, since private key operations still require bouncing through lsass. The perf gains are still measurable: around 100μs per op.

|         Method |        Job |    Toolchain |     Mean |   Error |  StdDev | Ratio |
|--------------- |----------- |------------- |---------:|--------:|--------:|------:|
|   Rsa_SignHash | Job-QLWHCM | cngkey_cache | 659.9 μs | 5.05 μs | 4.47 μs |  0.88 |
|   Rsa_SignHash | Job-OFFVFC |         main | 748.6 μs | 2.14 μs | 1.90 μs |  1.00 |
|                |            |              |          |         |         |       |
| Ecdsa_SignHash | Job-QLWHCM | cngkey_cache | 373.1 μs | 2.74 μs | 2.57 μs |  0.78 |
| Ecdsa_SignHash | Job-OFFVFC |         main | 479.1 μs | 3.02 μs | 2.52 μs |  1.00 |

GrabYourPitchforks committed:
Rename test class; fix incorrect SkipException usage

vcsjones commented:
> The perf gains are still measurable: around 100μs per op.

Sure. I’m not trying to argue against the change - we should take it regardless since it has already demonstrated some improvements elsewhere. Just noting that the internal partner is likely to get an even more significant performance improvement for .NET 8 once they start going through bcrypt.

bartonjs commented:
Really this shows that Windows should add similar caching.  But even if they did it'd save on the machinery of a P/Invoke.

vcsjones commented:
> Really this shows that Windows should add similar caching.

Well... the key size can change - before NCryptFinalize is called. We (mostly) don't have that problem here. But I'm not so sure how reasonable it would be for Windows to add caching with that in mind.

GrabYourPitchforks commented:
Looks like CI is a bit clogged. I'll come back in a few hours and see how things have progressed before merging.

IDisposable reviewed (approved):
:shipit:
